### PR TITLE
Flash v2 

### PR DESCRIPTION
**Pull request by hypnopump**
Uses flash attention v2. Works for any sequence length. 
* Flash attn v2 from pytorch for masked attention. 
* Custom flash attn v2 kernel for bias (triton-based, currently block size decided by autotuner, might need improvements to avoid recompilation). 
    * TF32 acceleration disabled by default, might need to check compute architecture and selectively enable it.   

WARNING! Pls install the custom fork from [trident](https://github.com/kakaobrain/trident) provided below as a zip file. You might need to install `triton-nightly` afterwards (maybe also `nvtx`) to get it working (see [triton repo](https://github.com/openai/triton/#quick-installation) for instructions). Check with `ipython; import trident as td`

[trident.zip](https://github.com/dptech-corp/Uni-Fold/files/13061121/trident.zip)

### CODE REVIEW

File: train_monomer_demo.sh [hunk 1]
```diff
@@ -2,15 +2,16 @@
 [ -z "${n_gpu}" ] && n_gpu=$(nvidia-smi -L | wc -l)
 export NCCL_ASYNC_ERROR_HANDLING=1
 export OMP_NUM_THREADS=1
-mkdir -p $1
+mkdir -p temp
```
Comment (hypnopump) [inline train_monomer_demo.sh:5]: once this proposal is validated, modifications to this file can be dropped; was just for easy testing of the code

File: unifold/modules/attentions.py [hunk 2]
```diff
@@ -52,34 +65,81 @@ def forward(
...
-        del attn, v
+        batch_dims = q.shape[:-2]
+
+        if bias is None and TORCH_FLASH:
+            # (b, n, i, (h dim_head)) -> (b, n, h, i, dim_head)
+            q = q.view(*q.shape[:-1], self.num_heads, self.head_dim).transpose(-2, -3).contiguous()
+            k = k.view(*k.shape[:-1], self.num_heads, self.head_dim).transpose(-2, -3).contiguous()
+            v = v.view(*v.shape[:-1], self.num_heads, self.head_dim).transpose(-2, -3)
+            # (b, n, h, i, dim_head), (b, n, h, j, dim_head) -> (b, n, h, i, dim_head)
+            o = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=mask)
```
Comment (guolinke) [inline unifold/modules/attentions.py:99]: How about its speed/memory compared with the one with softmax_dropout?
